### PR TITLE
feat(server): search unknown place

### PR DESCRIPTION
**Pull request by jforseth210**
This is a rough draft for allowing users to search for photos that specifically don't have a location (or at least a country) as requested in #8362. 

It updates the server to accept `null` as a value for country, which tells the database to search for assets with no country `exifInfo`. 

The frontend is updated to automatically include an "Unknown" option in the country combo box, which is passed to the server as `{"country":null}`. The filter list in the results is also updated to display "Unknown" if the value of a filter is `null`. 

I don't have a ton of experience contributing to FOSS, so this probably is a ways from merge ready. Let me know what I need to do differently. 

- I suspect there's something I need to update in `/open-api` but I'm not sure what.
- Removing the validators in `search.dto.ts` makes me a bit nervous. Is there anything that's expecting a string that I missed?
- I'd love some advice to make `selectedOption` easier to read
- I haven't touched the mobile app at all

### CODE REVIEW

**Comment by jforseth210** [comment]
I'm not really sure how to correctly update the types for `@immich/sdk`. Tried reading through the docs but still not getting it... Where's the best place to get a hand? Discord?

**Comment by danieldietzler** [comment]
> I'm not really sure how to correctly update the types for `@immich/sdk`. Tried reading through the docs but still not getting it... Where's the best place to get a hand? Discord?

`make open-api`. But yeah, generally joining our discord (https://discord.immich.app) is a good idea

File: server/src/dtos/search.dto.ts [hunk 1]
```diff
@@ -84,8 +84,6 @@ class BaseSearchDto {
   @Optional()
   state?: string;
 
-  @IsString()
-  @IsNotEmpty()
```
Comment (jrasm91) [inline server/src/dtos/search.dto.ts:88]: You need to add these back. The next line needs to be `@Optional({ nullable: true })` instead.

**Comment by jforseth210** [comment]
I think that last commit should fix those api issues. Thanks @danieldietzler and @jrasm91 for the help! What's next?

File: web/src/lib/components/shared-components/search-bar/search-location-section.svelte [hunk 2]
```diff
@@ -45,7 +49,16 @@
...
+  async function updateCities(country?: string | null, state?: string | null) {
+    if (country === null) {
+      cities = [];
+      return;
+    }
+    if (state == null) {
+      cities = [];
+      return;
+    }
+
```
Comment (jrasm91) [inline web/src/lib/components/shared-components/search-bar/search-location-section.svelte:61]: Combine these into one statement

Comment (jforseth210) [reply web/src/lib/components/shared-components/search-bar/search-location-section.svelte]: No problem!

**Review by danieldietzler — APPROVED** [review]

**Review by jrasm91 — APPROVED** [review]
Thanks for the PR! I made a few changes, but I think it is good to go now:
- Moved the logic to return "null" options to the server. This means they are only shown in the UI if there are actually valid results (given other criteria) with null values.
- Updated the api to accept empty string or null. (The mobile app can't actually send null, and empty string makes the combobox types a lot simpler)
- Added some e2e tests for searching with empty/null
- Expanded the scope to include camera make, camera model, and camera lens.

**Review by danieldietzler — APPROVED** [review]
LGTM

File: server/src/repositories/metadata.repository.ts [hunk 3]
```diff
@@ -60,49 +60,42 @@ export class MetadataRepository implements IMetadataRepository {
...
   @GenerateSql({ params: [DummyValue.UUID] })
   async getCountries(userId: string): Promise<string[]> {
-    const entity = await this.exifRepository
+    const results = await this.exifRepository
       .createQueryBuilder('exif')
       .leftJoin('exif.asset', 'asset')
       .where('asset.ownerId = :userId', { userId })
-      .andWhere('exif.country IS NOT NULL')
-      .select('exif.country')
+      .select('exif.country', 'exif_country')
```
Comment (mertalev) [inline server/src/repositories/metadata.repository.ts:67]: Is there a particular reason you set an alias here and make it a raw query? You can remove the alias and use `getMany` like the original, just removing the `filter` at the end.

File: server/src/utils/database.ts [hunk 4]
```diff
@@ -48,7 +48,13 @@ export function searchAssetBuilder(
       ? builder.leftJoinAndSelect(`${builder.alias}.exifInfo`, 'exifInfo')
       : builder.leftJoin(`${builder.alias}.exifInfo`, 'exifInfo');
 
-    builder.andWhere({ exifInfo });
+    for (const [key, value] of Object.entries(exifInfo)) {
+      if (value === null || value === '') {
+        builder.andWhere(`exifInfo.${key} IS NULL`);
```
Comment (mertalev) [inline server/src/utils/database.ts:53]: An empty string is not the same as null and shouldn't have this behavior.

File: web/src/lib/components/shared-components/search-bar/search-filter-box.svelte [hunk 5]
```diff
@@ -42,18 +42,22 @@
   const toStartOfDayDate = (dateString: string) => parseUtcDate(dateString)?.startOf('day').toISODate() || undefined;
   const dispatch = createEventDispatcher<{ search: SmartSearchDto | MetadataSearchDto }>();
 
+  function withoutNull<T>(value: T | null) {
```
Comment (mertalev) [inline web/src/lib/components/shared-components/search-bar/search-filter-box.svelte:45]: I'm not a huge fan of this. It adds a level of indirection just for a || or ?? operator. Just use the operator.

File: web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 6]
```diff
@@ -259,6 +259,8 @@
             {#await getPersonName(value) then personName}
               {personName}
             {/await}
+          {:else if value === null || value === ''}
```
Comment (mertalev) [inline web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte:262]:
```suggestion
          {:else if !value}
```

File: web/src/lib/components/shared-components/search-bar/search-camera-section.svelte [hunk 7]
```diff
@@ -26,13 +26,19 @@
       $type: SearchSuggestionType.CameraMake,
       model,
     });
+    if (filters.make && !makes.includes(filters.make)) {
+      filters.make = undefined;
+    }
```
Comment (mertalev) [inline web/src/lib/components/shared-components/search-bar/search-camera-section.svelte:31]: Why is this necessary?

File: server/src/dtos/search.dto.ts [hunk 8]
```diff
@@ -75,34 +75,28 @@ class BaseSearchDto {
...
 
   @IsString()
-  @IsNotEmpty()
-  @Optional()
-  model?: string;
+  @Optional({ nullable: true })
+  model?: string | null;
 
   @IsString()
-  @IsNotEmpty()
```
Comment (mertalev) [inline server/src/dtos/search.dto.ts:103]: I think the request should explicitly be null and not an empty string, so this doesn't need to be removed.

**Review by mertalev — COMMENTED** [review]
Thanks for working on this! I think some of the backend behavior is strange and surprising for the sake of letting the frontend send '' instead of null. I'd much prefer if you could make the frontend send null instead.

**Comment by mertalev** [comment]
> The mobile app can't actually send null

How can it not send null? That's so weird. In that case I'd rather handle the '' => null conversion as a transformer without polluting other code.

File: server/src/repositories/metadata.repository.ts (hunk 3 above)
Comment (jrasm91) [reply server/src/repositories/metadata.repository.ts:67]: Right, when I did that it only returned one result (United States) instead of two (United States, null).

File: web/src/lib/components/shared-components/search-bar/search-filter-box.svelte (hunk 5 above)
Comment (jrasm91) [reply web/src/lib/components/shared-components/search-bar/search-filter-box.svelte:45]: How do you propose doing it instead? `country: searchQuery.country || undefined` doesn't work since both empty string, 0 and null map to `undefined`. `country: searchQuery.country ?? undefined` also doesn't work, since it maps `null` to `undefined`.

File: web/src/lib/components/shared-components/search-bar/search-camera-section.svelte (hunk 7 above)
Comment (jrasm91) [reply web/src/lib/components/shared-components/search-bar/search-camera-section.svelte:31]: This just matches what was in the location filter file. If the current set filter isn't in the list of valid options it is discarded.

File: web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte (hunk 6 above)
Comment (jrasm91) [reply web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte:262]: That might work right now, it's not actually correct. It would make searching for a numeric value of 0 or `false` show the wrong value. `!value` impies "falsy", but what we actually want is "null or empty string".

File: server/src/dtos/search.dto.ts (hunk 8 above)
Comment (jrasm91) [reply server/src/dtos/search.dto.ts:103]: I could do that, at least for the web. But the mobile app can't actually send `null` values through the generated client. So either the mobile app doesn't get this functionality or we need to use an empty string in the api to indicate querying for null.

File: web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte (hunk 6 above)
Comment (mertalev) [reply web/src/routes/(user)/search/[[photos=photos]]/[[assetId=id]]/+page.svelte:262]: That's fair, it could lead to wrong behavior down the line.

File: web/src/lib/components/shared-components/search-bar/search-camera-section.svelte (hunk 7 above)
Comment (mertalev) [reply web/src/lib/components/shared-components/search-bar/search-camera-section.svelte:31]: Sure, but when can that actually happen?

File: server/src/repositories/metadata.repository.ts (hunk 3 above)
Comment (mertalev) [reply server/src/repositories/metadata.repository.ts:67]: Oh, maybe because you don't select the `id` field. That's fine, it can be a raw query. Maybe just make the alias camel case (or remove it).

File: web/src/lib/components/shared-components/search-bar/search-camera-section.svelte (hunk 7 above)
Comment (jrasm91) [reply web/src/lib/components/shared-components/search-bar/search-camera-section.svelte:31]: I think only by directly navigating with those params. Just seemed good to make them consistent, but maybe it isn't necessary

**Comment by jforseth210** [comment]
Is there anything I still need to do on this?

**Comment by mertalev** [comment]
The empty string / null handling is the main outstanding issue. It'd be great if you could add a '' => null transformer in the request DTO and remove the special-casing of empty strings for this elsewhere.

File: web/src/lib/components/shared-components/search-bar/search-camera-section.svelte (hunk 7 above)
Comment (mertalev) [reply web/src/lib/components/shared-components/search-bar/search-camera-section.svelte:32]: It can be removed IMO, but it's fine if you prefer to keep it.

File: e2e/src/api/specs/search.e2e-spec.ts [hunk 9]
```diff
@@ -108,7 +105,7 @@ describe('/search', () => {
     );
 
     await Promise.all(updates);
-    for (let i = 0; i < coordinates.length; i++) {
+    for (const [i] of coordinates.entries()) {
```
Comment (mertalev) [inline e2e/src/api/specs/search.e2e-spec.ts:108]: Interesting usage of `.entries()`.

**Review by mertalev — APPROVED** [review]

**Comment by maxime1992** [comment]
Hey, was there a regression on this?

Trying to filter my pics to find some with missing GPS metadata, I'm not able to:

image

And I've got some that have a location that are showing up :thinking: 

Do note though that the pics showing up had no GPS metadata initially. I added them on Immich but I'm not looking for pictures which don't have metadata attached onto the picture itself, I'm looking for pictures that don't have any GPS coordinate in Immich.

**Comment by jforseth210** [comment]
I'm not able to reproduce on my instance (1.124.2). Searched for photos with unknown location, added location to one, searched again, and that photo was no longer in the results.

**Comment by maxime1992** [comment]
Hi, I checked again and same issue for me. Also on `1.124.2`. 

I filter for `Unknown` country and for a certain camera model, I get some results with a location (on pics which did not have the location initally but it was manually added).

Is there a way for me to help debug this, maybe share logs or something else?

**Comment by duncaninnes** [comment]
Can I just point out that filtering on Unknown Country is not the same as filtering on Unknown Location.

I have many images which were taken at sea, and however Immich is calculating country, they are deemed to have Unknown Country.  So most of my boating/diving pictures come up when I search by Unknown Country.

Not sure if this means it would be better to provide a real search by Unknown Location, or provide better calculation of Country to include maritime water boundaries?  Or both?

But there is a gap here which leaves a lot of photos mis-attributed.
